### PR TITLE
Theia Backend

### PR DESCRIPTION
**Pull request by KevinK-9**
This pull request includes the communication from the frontend to the server

### CODE REVIEW

File: packages/deployment-server/src/deployment-artifacts/routes.ts [hunk 1]
```diff
@@ -23,10 +22,10 @@ export default function deploymentArtifactsRoutes (router: Router): void {
   router.post(
     '/deployment-artifacts',
     upload.single('file'),
-    validate<CreateArtifactResponseBody>({}),
     async (req, res, next) => {
       try {
         if (req.file == null) {
+          console.log(req.file)
```
Comment (WoH) [inline packages/deployment-server/src/deployment-artifacts/routes.ts:28]: Cleanup

File: packages/device-connector/src/index.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
+"use strict";
```
Comment (WoH) [inline packages/device-connector/src/index.js:1]: Ignore/delete build artifacts

File: packages/device-connector/test/rpc-sample.spec.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
+"use strict";
```
Comment (WoH) [inline packages/device-connector/test/rpc-sample.spec.js:1]: Ignore/delete build artifacts

File: packages/theia-extension/cdtcloud/configs/jest.config.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
+"use strict";
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/configs/jest.config.js:1]: Ignore/delete build artifacts

File: packages/grpc/package.json [hunk 5]
```diff
@@ -11,6 +11,7 @@
     "build:settings": "proto-loader-gen-types --keepCase --longs=String --enums=String --defaults --oneofs --grpcLib=@grpc/grpc-js -I proto/ --outDir=common/ cc/arduino/cli/settings/v1/settings.proto"
   },
   "dependencies": {
-    "@grpc/proto-loader": "^0.6.7"
+    "@grpc/proto-loader": "^0.6.7",
+    "tsc": "^2.0.3"
```
Comment (WoH) [inline packages/grpc/package.json:15]: why was this added?

File: packages/theia-extension/cdtcloud/src/browser/cdtcloud-command-contribution.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  CommandContribution,
+  CommandRegistry,
+} from "@theia/core/lib/common";
+import { inject, injectable } from "@theia/core/shared/inversify";
+import { HelloBackendWithClientService } from "../common/protocol";
+import { WorkspaceService } from "@theia/workspace/lib/browser/workspace-service";
+import { EditorManager } from '@theia/editor/lib/browser/editor-manager';
+
+
+const SayHelloViaBackendCommandWithCallBack: Command = {
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/browser/cdtcloud-command-contribution.ts:13]: Cleanup

File: packages/theia-extension/cdtcloud/src/browser/cdtcloud-command-contribution.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+@injectable()
+export class CdtCloudCommandContribution implements CommandContribution {
+  @inject(WorkspaceService)
+  protected readonly workspaceService: WorkspaceService;
+
+  @inject(EditorManager)
+  protected readonly editorManager: EditorManager;
+
+  constructor(
+    @inject(HelloBackendWithClientService)
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/browser/cdtcloud-command-contribution.ts:30]: Cleanup

File: packages/theia-extension/cdtcloud/src/browser/cdtcloud-command-contribution.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    private readonly helloBackendWithClientService: HelloBackendWithClientService,
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/browser/cdtcloud-command-contribution.ts:31]: Cleanup

File: packages/theia-extension/cdtcloud/src/browser/cdtcloud-command-contribution.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+      execute: () => {
+        // Get the currently opened path
+        console.log(this.workspaceService.workspace?.resource);
+        this.helloBackendWithClientService.greet().then((r) => console.log(r));
+      },
+    });
+    registry.registerCommand(CompileCommand, {
+      execute: () => {
+        const currentEditor = this.editorManager.currentEditor;
+        console.log("This is: " + currentEditor?.editor.document.uri);
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/browser/cdtcloud-command-contribution.ts:48]: Cleanup

File: packages/theia-extension/cdtcloud/src/browser/cdtcloud-command-contribution.ts [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        const sketchPath = currentEditor?.editor.document.uri
+        const workspaceService = this.workspaceService;
+        console.log(workspaceService.workspace?.name)
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/browser/cdtcloud-command-contribution.ts:51]: Cleanup

File: packages/theia-extension/cdtcloud/src/browser/cdtcloud-frontend-module.ts [hunk 11]
```diff
@@ -1,16 +1,69 @@
...
+import "../../src/browser/style/index.css";
+import { CdtCloudCommandContribution } from "./cdtcloud-command-contribution";
+import { CommandContribution } from "@theia/core";
+import {
+  BackendClient,
+  CompilationService,
+  COMPILATION_PATH,
+  DeviceTypeService,
+  DEVICE_TYPES_PATH,
+  HelloBackendWithClientService,
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/browser/cdtcloud-frontend-module.ts:20]: Cleanup

File: packages/theia-extension/cdtcloud/src/browser/cdtcloud-frontend-module.ts [hunk 12]
```diff
@@ -1,16 +1,69 @@
...
+    .inSingletonScope();
+
+  bind(CompilationService)
+  .toDynamicValue((ctx) => {
+    const connection = ctx.container.get(WebSocketConnectionProvider);
+    return connection.createProxy<CompilationService>(COMPILATION_PATH);
+  })
+  .inSingletonScope();
+
+  bind(HelloBackendWithClientService)
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/browser/cdtcloud-frontend-module.ts:52]: Cleanup

File: packages/theia-extension/cdtcloud/src/browser/cdtcloud-widget.tsx [hunk 13]
```diff
@@ -25,53 +40,75 @@ export class CdtcloudWidget extends ReactWidget {
...
-        headers: {
-          'Content-Type': 'application/json',
-        },
-        body: JSON.stringify(data),
-      });
+      this.deviceList = await this.deviceTypeService.getDeviceList()
+      this.options = this.deviceList.map(({ id, name }) => ({ label: name, value: id }))
+      this.update()
+    } catch (err) {
+      console.error(err)
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/browser/cdtcloud-widget.tsx:85]: Use logger

File: packages/theia-extension/cdtcloud/src/browser/cdtcloud-widget.tsx [hunk 14]
```diff
@@ -25,53 +40,75 @@ export class CdtcloudWidget extends ReactWidget {
...
+
   }
 
+  protected async deployOnBoard(board: any): Promise<void> {
+    const selectedBoard = this.deviceList.find(obj => {
+      return obj.id === board.value
+    })
 
+    const currentEditor = this.editorManager.currentEditor;
+    console.log("This is: " + currentEditor?.editor.document.uri);
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/browser/cdtcloud-widget.tsx:97]: Cleanup

File: packages/theia-extension/cdtcloud/src/browser/cdtcloud-widget.tsx [hunk 15]
```diff
@@ -25,53 +40,75 @@ export class CdtcloudWidget extends ReactWidget {
...
 
+    const workspaceService = this.workspaceService;
+    console.log(workspaceService)
+    const sketchPath = "" + workspaceService.workspace?.resource.path
+    console.log(sketchPath)
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/browser/cdtcloud-widget.tsx:102]: Cleanup

File: packages/theia-extension/cdtcloud/src/browser/cdtcloud-widget.tsx [hunk 16]
```diff
@@ -25,53 +40,75 @@ export class CdtcloudWidget extends ReactWidget {
...
 
+    if (sketchPath) {
+      this.compilationService.compile(selectedBoard.fqbn, board.value, "C:/Users/kevin/Documents/Arduino/Light_Project_1").then((result) => {
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/browser/cdtcloud-widget.tsx:105]: Remove hard coded path

File: packages/theia-extension/cdtcloud/src/browser/cdtcloud-widget.tsx [hunk 17]
```diff
@@ -25,53 +40,75 @@ export class CdtcloudWidget extends ReactWidget {
...
+        //Todo: Handle result
+        console.log(result)
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/browser/cdtcloud-widget.tsx:107]: Cleanup

File: packages/theia-extension/cdtcloud/src/browser/cdtcloud-widget.tsx [hunk 18]
```diff
@@ -25,53 +40,75 @@ export class CdtcloudWidget extends ReactWidget {
...
-      <button className='theia-button secondary' title='Display Message' onClick={_a => this.deployOnBoard()}>Deploy on Board</button>
+      <this.Selector />
     </div>
   }
 
-  protected async deployOnBoard(): Promise<void> {
-
-    const data = { deviceTypeId: '37dcb3ab-071d-43e1-935a-1e70403e4720', artifactUri: 'https://sanctum-dev.com/tests.bin' };
+  protected handleChange(option: { label: string; value: string }): void {
+    console.log(option)
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/browser/cdtcloud-widget.tsx:75]: Cleanup

File: packages/theia-extension/cdtcloud/src/browser/cdtcloud-widget.tsx [hunk 19]
```diff
@@ -25,53 +40,75 @@ export class CdtcloudWidget extends ReactWidget {
...
+  protected async deployOnBoard(board: any): Promise<void> {
+    const selectedBoard = this.deviceList.find(obj => {
+      return obj.id === board.value
+    })
 
+    const currentEditor = this.editorManager.currentEditor;
+    console.log("This is: " + currentEditor?.editor.document.uri);
 
+    const workspaceService = this.workspaceService;
+    console.log(workspaceService)
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/browser/cdtcloud-widget.tsx:100]: Cleanup

File: packages/theia-extension/cdtcloud/src/common/protocol.ts [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+
+export const CompilationService = Symbol("Compilation");
+export const COMPILATION_PATH = "/services/compilation";
+export interface CompilationService {
+  binaryFile: string;
+  artifactUri: string;
+  compile(fqbn:string, id:string, sketchPath:string): Promise<void>;
+}
+
+export const HelloBackendWithClientService = Symbol("BackendWithClient");
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/common/protocol.ts:18]: Cleanup

File: packages/theia-extension/cdtcloud/src/node/cdtcloud-backend-module.ts [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+  bind(ConnectionHandler)
+    .toDynamicValue(
+      (ctx) =>
+        new JsonRpcConnectionHandler(COMPILATION_PATH, () => {
+          return ctx.container.get<CompilationService>(CompilationService);
+        })
+    )
+    .inSingletonScope();
+
+  bind(HelloBackendWithClientService)
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/node/cdtcloud-backend-module.ts:37]: Cleanup

File: packages/theia-extension/cdtcloud/src/node/compilation-service.test.ts [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+  beforeEach(async () => {
+    compilationServiceImpl = new CompilationServiceImpl();
+  });
+
+  it("should make 2 requests to server", async () => {
+    try {
+      await compilationServiceImpl.compile(
+        "arduino:avr:mega",
+        "e8665d51-4221-4d3c-b892-a672a2af37a2",
+        "C:/Users/kevin/Documents/Arduino/Light_Project_1"
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/node/compilation-service.test.ts:16]: Make this reproducible by checking in a fixture we can verify against.

Extract the compilation functionality so we can test it in isolation.

We can move the integration testing up until we can simulate devices.

File: packages/theia-extension/cdtcloud/src/node/compilation-service.ts [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+  artifactUri: string;
+
+  async compile(fqbn:string, id: string, sketchPath: string): Promise<void> {
+    const client = new RPCClient()
+      await client.init()
+      await client.createInstance()
+      await client.initInstance()
+
+      const buildPath = await client.getBuildPath(fqbn, sketchPath)
+      console.log(buildPath)
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/node/compilation-service.ts:24]: Cleanup

File: packages/theia-extension/cdtcloud/src/node/compilation-service.ts [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+          }
+      })
+
+      let form = new FormData();
+      const content = createReadStream(this.binaryFile)
+      form.append('file', content)
+      const formHeaders = form.getHeaders();
+
+      const uploadResponse =
+        await got.post<{artifactUri: string}>(`http://localhost:3001/deployment-artifacts`, {
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/node/compilation-service.ts:39]: Host from config

File: packages/theia-extension/cdtcloud/src/node/compilation-service.ts [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+          headers: {
+            ...formHeaders
+          },
+          body: form,
+          responseType: 'json'
+        })
+
+        console.log(uploadResponse.body.artifactUri)
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/node/compilation-service.ts:47]: Cleanup

File: packages/theia-extension/cdtcloud/src/node/compilation-service.ts [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+
+      const artifactUri = uploadResponse.body.artifactUri
+
+      await got.post(`http://localhost:3001/deployments`, {
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/node/compilation-service.ts:51]: From config

File: packages/theia-extension/cdtcloud/src/node/device-types-service.ts [hunk 27]
```diff
@@ -0,0 +1,15 @@
+import axios from "axios";
+import { injectable } from "inversify";
+import { DeviceTypeService } from "../common/protocol";
+
+@injectable()
+export class DeviceTypeServiceImpl implements DeviceTypeService {
+  //get (query parameter, deployable= true)
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/node/device-types-service.ts:7]: Cleanup

File: packages/theia-extension/cdtcloud/src/node/device-types-service.ts [hunk 28]
```diff
@@ -0,0 +1,15 @@
...
+  async getDeviceList(): Promise<any[]> {
+    const { data } = await axios.get<any[]>(
+      "http://localhost:3001/device-types",
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/node/device-types-service.ts:10]: From config

File: packages/theia-extension/cdtcloud/src/node/hello-backend-with-client-service.ts [hunk 29]
```diff
@@ -0,0 +1,25 @@
+import { injectable } from "@theia/core/shared/inversify";
+import {
+  BackendClient,
+  HelloBackendWithClientService,
+} from "../common/protocol";
+
+@injectable()
+export class HelloBackendWithClientServiceImpl
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/node/hello-backend-with-client-service.ts:8]: Cleanup

File: packages/theia-extension/cdtcloud/src/node/rpc-client.ts [hunk 30]
```diff
@@ -0,0 +1,135 @@
...
+  buildPath?: string
+
+  constructor (address: string = '127.0.0.1:50051') {
+    this.address = address
+  }
+
+  async init (): Promise<void> {
+    const address = this.address
+    const packageDefinition = protoLoader.loadSync(
+      join(__dirname, '../../../../grpc/proto/cc/arduino/cli/commands/v1/commands.proto'), {
```
Comment (WoH) [inline packages/theia-extension/cdtcloud/src/node/rpc-client.ts:25]: Reproducible?

**Review by WoH — CHANGES_REQUESTED** [review]
See comments, also remove/ignore log files please

**Review by WoH — CHANGES_REQUESTED** [review]

**Review by WoH — DISMISSED** [review]
As discussed, the fixes will be handled in a separate PR

**Review by Gozzim — CHANGES_REQUESTED** [review]

File: packages/theia-extension/cdtcloud/src/node/compilation-service.ts [hunk 31]
```diff
@@ -0,0 +1,59 @@
...
+    await client.initInstance()
+
+    const buildPath = await client.getBuildPath(fqbn, sketchPath)
+    const files = await readdir(buildPath)
+
+    files.forEach((file) => {
+      if (file.endsWith('.bin')) {
+        this.binaryFile = join(buildPath, file)
+      }
+    })
```
Comment (Gozzim) [inline packages/theia-extension/cdtcloud/src/node/compilation-service.ts:30]: 1) You can break once the right file was found
2) Compiling can give you different files with the same extension. Make sure you select the right one (not ending in with_bootloader.{bin|hex})
3) Files extensions depend on the arduino core type

Something like this **Not tested**
```suggestion
// Method to determine required extension
const getRequiredExtension = async (fqbn: string): Promise<string> => {
  if (fqbn.startsWith('arduino')) {
    const segments = fqbn.split(':')
    switch (segments[1]) {
      case 'avr':
        return '.hex'
      case 'sam':
        return '.bin'
      case 'mbed_edge':
      case 'mbed_nano':
      case 'mbed_nicla':
      case 'mbed_portenta':
      case 'mbed_rp2040':
      case 'megaavr':
      case 'nrf52':
      case 'samd':
      case 'mbed':
        throw new Error('No extension specified')
      default:
        throw new Error('Unknown arduino board core')
    }
  }
  throw new Error('Unknown device')
}

const buildPath = await client.getBuildPath(fqbn, sketchPath)
const files = await readdir(buildPath)
const ext = await getRequiredExtension(fqbn)

files.forEach((file) => {
  if (Path.extname(file) === ext && !file.endsWith(`with_bootloader${ext}`)) {
    this.binaryFile = join(buildPath, file)
    break
  }
}

```

File: packages/theia-extension/cdtcloud/src/node/compilation-service.ts [hunk 32]
```diff
@@ -0,0 +1,59 @@
...
+        headers: {
+          ...formHeaders
+        },
+        body: form,
+        responseType: 'json'
+      })
+
+    const artifactUri = uploadResponse.body.artifactUri
+
+    await got.post(`http://localhost:3001/deployments`, {
```
Comment (Gozzim) [inline packages/theia-extension/cdtcloud/src/node/compilation-service.ts:48]: make url configurable. Deployment-server will not be on localhost

**Review by Gozzim — CHANGES_REQUESTED** [review]

File: packages/theia-extension/cdtcloud/src/browser/cdtcloud-widget.tsx [hunk 33]
```diff
@@ -25,53 +40,65 @@ export class CdtcloudWidget extends ReactWidget {
...
+    let sketchPath = "" + workspaceService.workspace?.resource.path
+    sketchPath = sketchPath.substring(sketchPath.indexOf('/') + 1);
+    console.log(sketchPath)
+    if (sketchPath) {
+      this.compilationService.compile(selectedBoard.fqbn, board.value, sketchPath)
+    }
 
-}
+    this.messageService.info('Deployment Request send.');
+  }
```
Comment (Gozzim) [inline packages/theia-extension/cdtcloud/src/browser/cdtcloud-widget.tsx:103]: There are going to be problems with the paths. Arduino-Cli expects a total path. The removed root `/` causes issues here. Because of the specific types of `Uri` and `Path` in theia, we can't resolve this with the normal methods. We could import the `FileUri` namespace here to get the correct path.


```suggestion
    const sketchUri = this.workspaceService.workspace?.resource
        if (sketchUri === undefined) {
          throw new Error('No Sketch found')
        }

        const sketchPath = FileUri.fsPath(sketchUri)
        await this.compilationService.compile(selectedBoard.fqbn, board.value, sketchPath)

        await this.messageService.info('Deployment Request send.');
      }
```

File: packages/theia-extension/cdtcloud/src/browser/cdtcloud-widget.tsx [hunk 34]
```diff
@@ -4,18 +4,33 @@ import { injectable, postConstruct, inject } from '@theia/core/shared/inversify'
 import { AlertMessage } from '@theia/core/lib/browser/widgets/alert-message';
 import { ReactWidget } from '@theia/core/lib/browser/widgets/react-widget';
 import { MessageService } from '@theia/core';
-
+import { CompilationService, DeviceTypeService } from "../common/protocol";
+import { EditorManager } from '@theia/editor/lib/browser/editor-manager';
+import { WorkspaceService } from '@theia/workspace/lib/browser/workspace-service';
```
Comment (Gozzim) [inline packages/theia-extension/cdtcloud/src/browser/cdtcloud-widget.tsx:10]: Import for comment below

```suggestion
import { FileUri } from "@theia/core/lib/node/file-uri";

 ```

File: packages/theia-extension/cdtcloud/src/node/compilation-service.ts [hunk 35]
```diff
@@ -0,0 +1,59 @@
+import { injectable } from "inversify";
+import { CompilationService } from "../common/protocol";
+import { readdir } from 'fs/promises';
+import { createReadStream } from 'fs';
+import { RPCClient } from "./rpc-client";
+import 'reflect-metadata';
+import FormData from 'form-data';
+import { join } from 'path';
+import got from "got";
+
```
Comment (Gozzim) [inline packages/theia-extension/cdtcloud/src/node/compilation-service.ts:10]: Import for comment below

```suggestion
import * as Path from "path";

 ```

**Review by Gozzim — CHANGES_REQUESTED** [review]
Discovered an issue caused by file paths during extended testing

**Comment by Gozzim** [comment]
Included in #16
